### DISCUSSION TITLE
naive compute_layer definition fails

### DISCUSSION THREAD

**Discussion by EvaMaeRey** [ggplot2-extenders/ggplot-extension-club#16] [Q&A] [answered]
I thought I'd document this naive stab at defining compute_layer which throws an error - follows compute_group and compute_panel construction.  I haven't explored further.

``` r
library(tidyverse)

ggplot(cars) + 
  aes(speed, dist) + 
  geom_point(size = .15) ->
base; base
```

![](https://i.imgur.com/bo3WXtX.png)

``` r
  

compute_mean <- function(data, scales){
  
  data %>% 
    summarise(x = mean(x),
              y = mean(y))
  
}

########  groupwise computation via compute group ####
StatMeangroup <- ggproto(`_class` = "StatMeangroup",
                         `_inherit` = Stat,
                         compute_group = compute_mean)



base +
  layer(geom = "point", 
        stat = StatMeangroup, 
        position = "identity") + 
  aes(color = speed > 15)
```

![](https://i.imgur.com/hKhzs6K.png)

``` r


########  panel-wise computation by defining compute panel #####
StatMeanpanel <- ggproto(`_class` = "StatMeanpanel",
                         `_inherit` = Stat,
                         compute_panel = compute_mean)

base +
  layer(geom = "point", 
        stat = StatMeanpanel, 
        position = "identity") + 
  aes(color = speed > 15)
```

![](https://i.imgur.com/CaMk5c4.png)

``` r

last_plot() +
  facet_wrap(facets = vars(speed > 15))
```

![](https://i.imgur.com/0TUK5AZ.png)

``` r


########  layer wise computation via compute_layer??? ####
StatMeanlayer <- ggproto(`_class` = "StatMeanlayer",
                         `_inherit` = Stat,
                         compute_layer = compute_mean)


# expect a point in each facet that computes the mean for the entire layer. 
base +
  layer(geom = "point", 
        stat = StatMeanlayer, 
        position = "identity") + 
  aes(color = speed > 15) +
  facet_wrap(facets = vars(speed > 15))
#> Error:
#> ! Problem while computing stat.
#> ℹ Error occurred in the 2nd layer.
#> Caused by error in `compute_layer()`:
#> ! unused argument (<environment>)
```

<sup>Created on 2024-05-30 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Comment by teunbrand** [comment, marked as answer]
The 3 different orders of `compute_*()` methods take different arguments.
Largely, the `compute_layer()` and `compute_panel()` methods serve to preserve as much static mappings as they reasonably can.
If you define your compute function as follows, I think it should work:

```r
compute_mean <- function(data, ...){
  data %>% 
    summarise(x = mean(x),
              y = mean(y))
}
```

However, as you're implicitly dropping the `PANEL` variable at the `compute_layer()` level, the layer will not be rendered.

**Reply by EvaMaeRey** [reply]
'However, as you're implicitly dropping the PANEL variable at the compute_layer() level, the layer will not be rendered.'  That makes sense.

**Comment by EvaMaeRey** [comment]
A plot rendered with the '...'. No error!  

But not the mean point did not render and I think because the computation (summarize w/ no grouping) drops PANEL variable, as you mention.  

Below, compute_mean records the Panels and adds them back after the means are computed.  Also tried grouping by PANEL before summarize to preserve the column, but then the means reflect only the data in each facet, which wasn't the behavior that I was interested in.   

``` r
library(tidyverse)

compute_mean <- function(data, ...){
  
  data %>% 
    distinct(PANEL) ->
  panel_df
  
  data %>% 
    summarise(x = mean(x),
              y = mean(y)) %>% 
    crossing(panel_df)
      
  
}

########  groupwise computation via compute group ####
StatMeanlayer <- ggproto(`_class` = "StatMeanlayer",
                         `_inherit` = Stat,
                         compute_layer = compute_mean)

ggplot(cars) + 
  aes(speed, dist) + 
  geom_point(size = .15) +
  layer(geom = "point", 
        stat = StatMeanlayer, 
        position = "identity")
```

![](https://i.imgur.com/z2UIT3h.png)

``` r

last_plot() +
  facet_wrap(facets = vars(speed > 10))
```

![](https://i.imgur.com/lMc7QbD.png)

``` r

layer_data(i = 2)
#> # A tibble: 2 × 9
#>       x     y PANEL shape colour  size fill  alpha stroke
#>   <dbl> <dbl> <fct> <dbl> <chr>  <dbl> <lgl> <lgl>  <dbl>
#> 1  15.4  43.0 1        19 black    1.5 NA    NA       0.5
#> 2  15.4  43.0 2        19 black    1.5 NA    NA       0.5
```

<sup>Created on 2024-05-31 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Reply by EvaMaeRey** [reply]
I guess this isn't very conventional, so will be interested to look at other approaches and how others define.  to do.

**Comment by EvaMaeRey** [comment]
More compute layer adventures...A stat that will help you annotate panel 2. 

``` r
library(tidyverse)

compute_annotation2_spot <- function(data, ...){

  data %>% 
    distinct(PANEL) ->
  panel_df
  
  data %>% 
    summarise(x = quantile(range(x), .05),
              y = quantile(range(y), .95)) %>% 
    crossing(panel_df) %>% 
    mutate(hjust = 0) %>% 
    mutate(vjust = 1) %>% 
    filter(PANEL == 2)
      
  
}

########  groupwise computation via compute group ####
StatAnnotateLayerPanel2 <- ggproto(`_class` = "StatAnnotateLayerPanel2",
                         `_inherit` = Stat,
                         compute_layer = compute_annotation2_spot)


ggplot(cars) + 
  aes(speed, dist) + 
  geom_point(size = .15) +
  layer(geom = "text", 
        stat = StatAnnotateLayerPanel2, 
        position = "identity",
        params = list(label = "Let's talk about\nthe 'TRUE' case")) +
  facet_wrap(facets = vars(speed > 10))
```

![](https://i.imgur.com/6vEtitS.png)

<sup>Created on 2024-06-21 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Reply by teunbrand** [reply]
Aside from whether it should be a stat, the problem of making panel specific annotation is interesting. For your stat specifically, shouldn't the panel number be a user-provided variable rather than fixed? :p The `quantile(range(x), ...)` thing seems strange, and probably only works because the default `type = 7`. Anyway you could probably skip that entirely by using `I(x)` for where to place it.

The following is similar to what I used to do quite a lot, which is to only have the relevant level in the `data` argument.

```r
ggplot(mpg, aes(displ, hwy)) +
  geom_point() +
  geom_text(
    data = data.frame(year = 2008),
    aes(x = I(0.2), y = I(0.9), label = "Hello world")
  ) +
  facet_wrap(~year)
```

Knowing a little more about the guts of ggplot2, you can totally hack in which panel something goes by enabling `PANEL` as a valid aesthetic and then mapping an integer-factor to that.

```r
update_geom_defaults("text", list(PANEL = NULL))

ggplot(mpg, aes(displ, hwy)) +
  geom_point() +
  annotate(
    "text", x = I(0.2), y = I(0.9), 
    label = "Hello world", PANEL = factor(2, levels = 1:2)
  ) +
  facet_wrap(~year)

```

**Reply by EvaMaeRey** [reply]
Panel specific annotation is interesting. Mostly I learned that I don't have a very good handle on specifying compute_layer - i tried to do a variable input for panel, but only managed hard coding specific panel.  

I()s do seem like the way to go

I started here, which is kind of doing the same things as happen in your geom_text strategy. 

``` r
stamp_text_facet <- function(facet, which, text = "sometext", 
                             x = I(.05), y = I(.95), hjust = 0,...){
  
    geom_text(data = . %>% 
              distinct({{facet}}) %>% 
              mutate(a_note = ifelse({{facet}} %in% which, 
                                    text, NA)),
            aes(label = a_note), x = x, y = y,
            hjust = hjust, ...)
  
  
}

palmerpenguins::penguins %>% 
  ggplot() + 
  facet_wrap(~species) +
  stamp_text_facet(facet = species, 
                   "Adelie", 
                   "As you can see...")
```

![](https://i.imgur.com/a9PuKfd.png)

<sup>Created on 2024-06-21 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Reply by teunbrand** [reply]
It seems like you're just missing the `params` object that gets passed to `compute_layer()`, which holds any parameters passed to the stat. If it isn't an argument to any `compute_*()` function, you should put it in the `extra_params` field.

**Reply by teunbrand** [reply]
Also this problem triggered the following function which I think could be pretty cool, especially since facet specification is relatively flexible.

``` r
library(ggplot2)

at_panel <- function(layer, expr) {
  expr <- rlang::enquo(expr)
  old_geom <- layer$geom
  new_geom <- ggproto(
    NULL, old_geom,
    draw_layer = function(self, data, params, layout, coord) {
      panels <- as.logical(rlang::eval_tidy(expr, layout$layout))
      panels <- rep_len(panels, nrow(layout$layout))
      panels <- layout$layout$PANEL[panels]
      data <- data[data$PANEL %in% panels, , drop = FALSE]
      ggproto_parent(old_geom, self)$draw_layer(data, params, layout, coord)
    }
  )
  ggproto(NULL, layer, geom = new_geom)
}

p <- ggplot(mpg, aes(displ, hwy)) +
  geom_point() +
  facet_wrap(~class)

annotation <- annotate("text", x = I(0.8), y = I(0.9), label = "Hello world")

p + at_panel(annotation, PANEL == 2)
```

![](https://i.imgur.com/DI7MKXP.png)

``` r
p + at_panel(annotation, COL == 2 & ROW == 2)
```

![](https://i.imgur.com/aY6im8S.png)

``` r
p + at_panel(annotation, class == "subcompact")
```

![](https://i.imgur.com/r91porO.png)

<sup>Created on 2024-06-21 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Reply by yjunechoe** [reply]
Is it bad that I'm really liking this style of ad-hoc ggproto extensions? 😆

The only concern I had with this approach was introducing potential naming conflicts with other ggprotos but I see you can also set the name to `NULL`. Am I right in thinking that this means it won't be in the ggproto "search path", and that you're instantiating a *new* Geom object every time you call `at_panel()`?

**Reply by teunbrand** [reply]
I usually use unnamed ggproto classes as a type of 'copy on modify' mechanic on the fly (they innately have reference semantics). But yes, it subclasses the geom (and layer instance) every time it is called.

I don't think there is any search path mirrored in ggproto class names, it is all class object names. For example, the `check_subclass()` function looks for a `GeomPoint` object, rather than a classed object with `"GeomPoint"` as the class name.

**Reply by teunbrand** [reply]
I'm going to self-plagiarise and put this in ggh4x. It is exactly the kind of package to host these kind of ad-hoc hacks :')

**Reply by EvaMaeRey** [reply]
The multiple ways of specifying which panel to annotate is pretty sweet.  

Combine with identity x and y - it's almost *too* easy to do annotation now!  I think the principled/production/careful way to define where is using conditioning via faceting variable (so your annotation matches content), or PANEL == 1 (so your generic annotation always shows up) but it's ggh4x so...

**Reply by teunbrand** [reply]
The issue I have with only allowing panel selection with facetting variables, is that you can use computation in specifying facets and `year * 2 + 5` is not a nice variable to work with. It would also be awkward with `margins = TRUE` that doesn't correspond to some particular level.

**Comment by EvaMaeRey** [comment]
`at_panel()` 🤯 

so revised the initial stat idea with I()s: `stamp_text_panel_two`.  generalizing with params later.  Feels useable.


``` r
library(tidyverse)

compute_annotation2_spot <- function(data, ...){

  data %>% 
    distinct(PANEL) %>%
    filter(PANEL == 2)
      
  
}

########  groupwise computation via compute group ####
StatAnnotateLayerPanel2 <- ggproto(`_class` = "StatAnnotateLayerPanel2",
                         `_inherit` = Stat,
                         compute_layer = compute_annotation2_spot)

stamp_text_panel_two <- function(label = "Let's talk about\nthe 'TRUE' case",
                      x = I(.025),
                      y = I(.975),
                      hjust = 0,
                      vjust = 1){
  
  layer(geom = "text", 
        stat = StatAnnotateLayerPanel2, 
        position = "identity",
        params = list(label = label,
                      x = x,
                      y = y,
                      hjust = hjust,
                      vjust = vjust)) 
  
} 

ggplot(cars) + 
  aes(speed, dist) + 
  geom_point(size = .15) +
  facet_wrap(facets = vars(speed > 10)) +
  stamp_text_panel_two("Panel two is\nthe best panel!")
```

![](https://i.imgur.com/5hBXwcR.png)

<sup>Created on 2024-06-21 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Comment by EvaMaeRey** [comment]
layout argument 🎉cross reference:  https://github.com/tidyverse/ggplot2/pull/6336
